gobblin-service/src/main/java/org/apache/gobblin/service/modules/template_catalog/UpdatingFSFlowTemplateCatalog.java

```diff
@@ -0,0 +1,108 @@
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.apache.hadoop.fs.Path;
+
+import com.typesafe.config.Config;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.gobblin.runtime.api.JobTemplate;
+import org.apache.gobblin.runtime.api.SpecNotFoundException;
+import org.apache.gobblin.service.modules.template.FlowTemplate;
+import org.apache.gobblin.util.filesystem.PathAlterationListener;
+import org.apache.gobblin.util.filesystem.PathAlterationListenerAdaptor;
+
+
+/**
+ * {@link FSFlowTemplateCatalog} that keeps a cache of flow and job templates. It has a
+ * {@link org.apache.gobblin.util.filesystem.PathAlterationListener} on the root path, and clears the cache when a change
+ * is detected so that the next call to {@link #getFlowTemplate(URI)} will use the updated files.
+ */
+@Slf4j
+public class UpdatingFSFlowTemplateCatalog extends FSFlowTemplateCatalog {
+  private static Map<URI, FlowTemplate> flowTemplateMap = new ConcurrentHashMap<>();
+  private static Map<URI, List<JobTemplate>> jobTemplateMap = new ConcurrentHashMap<>();
+
+  public UpdatingFSFlowTemplateCatalog(Config sysConfig) throws IOException {
+    super(sysConfig);
+  }
+
+  @Override
+  public FlowTemplate getFlowTemplate(URI flowTemplateDirURI)
+      throws SpecNotFoundException, JobTemplate.TemplateException, IOException, URISyntaxException {
+    if (!flowTemplateMap.containsKey(flowTemplateDirURI)) {
+      flowTemplateMap.put(flowTemplateDirURI, super.getFlowTemplate(flowTemplateDirURI));
+    }
+    return flowTemplateMap.get(flowTemplateDirURI);
+  }
+
+  @Override
+  public List<JobTemplate> getJobTemplatesForFlow(URI flowTemplateDirURI)
+      throws IOException, SpecNotFoundException, JobTemplate.TemplateException, URISyntaxException {
+    if (!jobTemplateMap.containsKey(flowTemplateDirURI)) {
+      jobTemplateMap.put(flowTemplateDirURI, super.getJobTemplatesForFlow(flowTemplateDirURI));
+    }
+    return jobTemplateMap.get(flowTemplateDirURI);
+  }
+
+  @Override
+  protected PathAlterationListener getListener() {
+    return new FlowCatalogPathAlterationListener();
+  }
+
+  @Override
+  protected void startUp() throws IOException {
+    if (this.pathAlterationDetector != null) {
+      this.pathAlterationDetector.start();
+    }
+  }
+
+  /**
+   * Clear cached templates so they will be reloaded next time {@link #getFlowTemplate(URI)} is called.
+   */
+  public void clearTemplates() {
+    log.info("Change detected, reloading flow templates.");
+    flowTemplateMap.clear();
+    jobTemplateMap.clear();
+  }
+
+  /**
+   * {@link org.apache.gobblin.util.filesystem.PathAlterationListener} that clears job template cache if a file is
+   * created or updated.
+   */
+  private class FlowCatalogPathAlterationListener extends PathAlterationListenerAdaptor {
+    @Override
+    public void onFileCreate(Path path) {
+      clearTemplates();
+    }
+
+    @Override
+    public void onFileChange(Path path) {
+      clearTemplates();
```

Comment:
Should it be a invalidation of an entry in cache?

Reply:
There are cases where, for example, a job template could get updated and multiple flow templates are pointing to that job template, so we would have to figure out which flows used that job template and invalidate those as well. Due to this I think it's easier to just clear entire cache.